lib/Authentication.ts

```diff
@@ -49,6 +52,150 @@ export default class Authentication {
+  public async signAuthenticationRequest (request: AuthenticationRequest): Promise<string> {
+    if (request.response_type !== 'id_token' || request.scope !== 'openid') {
+      throw new Error('Authentication Request not formed correctly');
+    }
+    const requesterDid = request.iss;
+    let key: PrivateKey | undefined;
+    key = this.getKey(requesterDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${requesterDid}`);
+    }
+
+    const token = new JwsToken(request, this.factory);
+    return token.sign(key);
+  }
+
+  /**
+   * Verifies signature on request and returns AuthenticationRequest.
+   * @param request Authentiation Request as a buffer or string.
+   */
+  public async verifyAuthenticationRequest (request: Buffer | string): Promise<AuthenticationRequest> {
+    // identity function in place of later signature verification
+    let jwsToken: JwsToken;
+    if (request instanceof Buffer) {
+      jwsToken = new JwsToken(request.toString(), this.factory);
+    } else {
+      jwsToken = new JwsToken(request, this.factory);
+    }
+    const keyId = jwsToken.getHeader().kid;
+    const keyDid = DidDocument.getDidFromKeyId(keyId);
+    const results = await this.resolver.resolve(keyDid);
+    const didPublicKey = results.didDocument.getPublicKey(keyId);
+    if (!didPublicKey) {
+      throw new Error('Could not find public key');
+    }
+    const publicKey = this.factory.constructPublicKey(didPublicKey);
+    const content = await jwsToken.verifySignature(publicKey);
+
+    const verifiedRequest: AuthenticationRequest = JSON.parse(content);
+    if (verifiedRequest.iss !== keyDid) {
+      throw new Error('Signing DID does not match issuer');
+    }
+    return verifiedRequest;
+  }
+
+  /**
+   * Given a challenge, forms a signed response using a given DID that expires at expiration, or a default expiration.
+   * @param authRequest Challenge to respond to
+   * @param responseDid The DID to respond with
+   * @param claims Claims that the requester asked for
+   * @param expiration optional expiration datetime of the response
+   */
+  public async formAuthenticationResponse (authRequest: AuthenticationRequest, responseDid: string, claims: any, expiration?: Date): Promise<string> {
+    let key: PrivateKey | undefined;
+    key = this.getKey(responseDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${responseDid}`);
+    }
+
+    const publicKey: PublicKey = key.getPublicKey();
+
+    // milliseconds to seconds
+    const milliseconds = 1000;
+    if (!expiration) {
+      const expirationTimeOffsetInMinutes = 5;
+      expiration = new Date(Date.now() + milliseconds * 60 * expirationTimeOffsetInMinutes); // 5 minutes from now
+    }
+    const iat = Math.floor(Date.now() / milliseconds); // ms to seconds
+    const response: AuthenticationResponse = {
+      iss: 'https://self-issued.me',
+      sub: responseDid,
+      aud: authRequest.client_id,
+      nonce: authRequest.nonce,

```

Comment:
What happens when the request did not provide a response? Does an empty nonce prpoerty get added to the token? If so it would be better to only add nonce property if one specified in the request. Having an empty nonce property might have unintended side-effects